### PR TITLE
iDeep: F.relu

### PR DESCRIPTION
**Pull request by niboshi**
iDeep support in F.relu (Separated from #3865. Based on #4276.)

### CODE REVIEW

File: chainer/functions/activation/relu.py [hunk 1]
```diff
@@ -17,36 +18,62 @@ class ReLU(function_node.FunctionNode):
...
             in_types[0].dtype.kind == 'f',
         )
 
-    def forward_cpu(self, x):
+    def forward_cpu(self, inputs):
+        if (intel64.should_use_ideep('>=auto')
+                and intel64.inputs_all_ready(inputs)):
+
+            # iDeep implementation
+            # TODO(iDeep): Support arbitrary dimension
```
Comment (kmaehashi) [inline chainer/functions/activation/relu.py]: Is this a TODO by us, or iDeep limitation?
(`inputs_all_ready` says `ideep supports array dimension in either 2 or 4 only.`)

In the latter case, maybe we don't need a TODO line in our code.

Comment (niboshi) [reply chainer/functions/activation/relu.py]: This comment was present in Intel's original implementation.
I think it's obsolete, so I removed it.

File: chainer/functions/activation/relu.py [hunk 2]
```diff
@@ -17,36 +18,62 @@ class ReLU(function_node.FunctionNode):
...
+            x, = self.get_retained_inputs()
+            return ReLUGradIdeep(x, y).apply((gy,))
         if chainer.should_use_cudnn('==always') and self._use_cudnn:
-            x = self.get_retained_inputs()[0]
-            return ReLUGrad3(x, y).apply((gy[0],))
-        else:
-            return ReLUGrad2(y).apply((gy[0],))
+            # cuDNN implementatioin
```
[the remark this answers is not in the document]
Comment (niboshi) [reply chainer/functions/activation/relu.py]: Fixed.

File: chainer/functions/activation/relu.py [hunk 3]
```diff
@@ -17,36 +18,62 @@ class ReLU(function_node.FunctionNode):
...
+            x, = self.get_retained_inputs()
+            return ReLUGradCudnn(x, y).apply((gy,))
+        # Generic implementaion
```
[the remark this answers is not in the document]
Comment (niboshi) [reply chainer/functions/activation/relu.py]: Fixed.

**Comment by kmaehashi** [comment]
Jenkins, test this please.

**Comment by kmaehashi** [comment]
This PR depends on #4278

**Comment by kmaehashi** [comment]
Jenkins, test this please.

**Comment by kmaehashi** [comment]
LGTM!

**Pull request merged**